returnn/tf/layers/basic.py

```diff
@@ -913,37 +913,204 @@ def transform_config_dict(cls, d, network, get_layer):
+    else:
+      return len(common_axes_input) + len(position_axes) + input_axes.index(input_axis)
 
   @classmethod
-  def get_out_data_from_opts(cls, name, sources, axis, **kwargs):
+  def translate_position_axis(
+    cls, position_axis, common_axes_input, common_axes_position, position_axes):
+    """
+    :param int position_axis: batch axis of position_data
+    :param list[int] common_axes_input:
+    :param list[int] common_axes_position:
+    :param list[int] position_axes:
+    :rtype: int
+    :return: batch axis of output
+    """
+    if position_axis in common_axes_position:
+      return common_axes_position.index(position_axis)
+    else:
+      return len(common_axes_input) + position_axes.index(position_axis)
+
+  def get_dep_layers(self):
+    """
+    :rtype: list[LayerBase]
+    """
+    if isinstance(self.position, LayerBase):
+      return super(GatherLayer, self).get_dep_layers() + [self.position]
+    else:
+      return super(GatherLayer, self).get_dep_layers()
+
+  @classmethod
+  def get_out_data_from_opts(cls, name, sources, position, axis, **kwargs):
     """
     :param str name:
     :param list[LayerBase] sources:
+    :param LayerBase|int position:
     :param str axis:
     :rtype: Data
     """
-    out = get_concat_sources_data_template(sources, name="%s_output" % name)
-    axis = out.get_axis_from_description(axis, allow_int=False)
-    return out.copy_template_excluding_axis(axis)
+    input_data = get_concat_sources_data_template(sources)
+    gather_axis = input_data.get_axis_from_description(axis, allow_int=False)
+
+    if isinstance(position, int):
+      position = ConstantLayer.get_out_data_from_opts(
+        name="constant_position", network=kwargs["network"], value=position, dtype="int32")
+    position_data = position.output.copy_template()
+    assert position_data.dtype in ["int32", "int64"]
+
+    # determine all common axes of input_data and position_data
+    common_axes_input, common_axes_position, input_axes, position_axes = (
+      cls._get_common_input_position_axes(input_data, position_data, gather_axis))
+
+    # (BatchAxes.., PositionAxes.., InputAxes..)
+    shape = (
+      [input_data.batch_shape[ax] for ax in common_axes_input if not ax == input_data.batch_dim_axis] +
+      [position_data.batch_shape[ax] for ax in position_axes if not ax == position_data.batch_dim_axis] +
+      [input_data.batch_shape[ax] for ax in input_axes if not ax == input_data.batch_dim_axis])
+    out_type = input_data.get_kwargs()
+    out_type["name"] = "%s_output" % name
+    out_type["shape"] = shape
+    out_type["beam"] = SearchBeam.get_combined_beam(input_data.beam, position_data.beam)
+
+    # Take axes from input_data if they exist there, otherwise from position_data
+    for axis_kind in ["batch_dim_axis", "time_dim_axis", "feature_dim_axis"]:
+      input_axis, position_axis = getattr(input_data, axis_kind), getattr(position_data, axis_kind)
```

Comment:
Actually, there is some special logic for `feature_dim_axis`, which you should also follow.
In that case, you should check `input_data.feature_dim_axis_or_unspecified is NotSpecified`, and if so, just do not set it explicitly to an integer. 
(This is somewhat not nice here, as this needs a special case... But whether it is not specified, i.e. automatically derived or explicitly specified can change the behavior in some cases, e.g. when dropping some dimension, when reordering the axes, etc.)

Reply:
@albertz Ah okay, thanks, I didn't know about this. Can you check if it's correct now? I'm still unsure.